File: app/src/main/java/me/zhanghai/android/files/filelist/ConfirmDeleteFilesDialogFragment.java
/*
 * Copyright (c) 2018 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.filelist;

import android.app.Dialog;
import android.os.Bundle;

import java.util.ArrayList;
import java.util.LinkedHashSet;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatDialogFragment;
import androidx.fragment.app.Fragment;
import me.zhanghai.android.files.R;
import me.zhanghai.android.files.util.CollectionUtils;
import me.zhanghai.android.files.util.FragmentUtils;
import me.zhanghai.java.functional.Functional;

public class ConfirmDeleteFilesDialogFragment extends AppCompatDialogFragment {

    private static final String KEY_PREFIX = ConfirmDeleteFilesDialogFragment.class.getName() + '.';

    private static final String EXTRA_FILES = KEY_PREFIX + "FILES";

    @NonNull
    private LinkedHashSet<FileItem> mExtraFiles;

    @NonNull
    private static ConfirmDeleteFilesDialogFragment newInstance(
            @NonNull LinkedHashSet<FileItem> files) {
        //noinspection deprecation
        ConfirmDeleteFilesDialogFragment fragment = new ConfirmDeleteFilesDialogFragment();
        FragmentUtils.getArgumentsBuilder(fragment)
                .putParcelableArrayList(EXTRA_FILES, new ArrayList<>(files));
        return fragment;
    }

    public static void show(@NonNull LinkedHashSet<FileItem> files, @NonNull Fragment fragment) {
        ConfirmDeleteFilesDialogFragment.newInstance(files)
                .show(fragment.getChildFragmentManager(), null);
    }

    /**
     * @deprecated Use {@link #newInstance(LinkedHashSet)} instead.
     */
    public ConfirmDeleteFilesDialogFragment() {}

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        mExtraFiles = new LinkedHashSet<>(getArguments().getParcelableArrayList(EXTRA_FILES));
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
        String message;
        if (mExtraFiles.size() == 1) {
            FileItem file = CollectionUtils.first(mExtraFiles);
            int messageRes = file.getAttributesNoFollowLinks().isDirectory() ?
                    R.string.file_delete_message_directory_format
                    : R.string.file_delete_message_file_format;
            message = getString(messageRes, FileUtils.getName(file));
        } else {
            boolean allDirectories = Functional.every(mExtraFiles, file ->
                    file.getAttributesNoFollowLinks().isDirectory());
            boolean allFiles = Functional.every(mExtraFiles, file ->
                    !file.getAttributesNoFollowLinks().isDirectory());
            int messageRes = allDirectories ?
                    R.string.file_delete_message_multiple_directories_format
                    : allFiles ? R.string.file_delete_message_multiple_files_format
                    : R.string.file_delete_message_multiple_mixed_format;
            message = getString(messageRes, mExtraFiles.size());
        }
        return new AlertDialog.Builder(requireContext(), getTheme())
                .setMessage(message)
                .setPositiveButton(android.R.string.ok, (dialog, which) -> getListener()
                        .deleteFiles(mExtraFiles))
                .setNegativeButton(android.R.string.cancel, null)
                .create();
    }

    @NonNull
    private Listener getListener() {
        return (Listener) requireParentFragment();
    }

    public interface Listener {
        void deleteFiles(@NonNull LinkedHashSet<FileItem> files);
    }
}

File: app/src/main/java/me/zhanghai/android/files/filelist/ConfirmDeleteFilesDialogFragment.kt
/*
 * Copyright (c) 2018 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.filelist

import android.app.Dialog
import android.os.Bundle
import androidx.appcompat.app.AppCompatDialogFragment
import androidx.fragment.app.Fragment
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import kotlinx.parcelize.Parcelize
import me.zhanghai.android.files.R
import me.zhanghai.android.files.util.ParcelableArgs
import me.zhanghai.android.files.util.args
import me.zhanghai.android.files.util.getQuantityString
import me.zhanghai.android.files.util.putArgs
import me.zhanghai.android.files.util.show

class ConfirmDeleteFilesDialogFragment : AppCompatDialogFragment() {
    private val args by args<Args>()

    private val listener: Listener
        get() = requireParentFragment() as Listener

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val files = args.files
        val message = if (files.size == 1) {
            val file = files.single()
            val messageRes = if (file.attributesNoFollowLinks.isDirectory) {
                R.string.file_delete_message_directory_format
            } else {
                R.string.file_delete_message_file_format
            }
            getString(messageRes, file.name)
        } else {
            val allDirectories = files.all { it.attributesNoFollowLinks.isDirectory }
            val allFiles = files.none { it.attributesNoFollowLinks.isDirectory }
            val messageRes = when {
                allDirectories -> R.plurals.file_delete_message_multiple_directories_format
                allFiles -> R.plurals.file_delete_message_multiple_files_format
                else -> R.plurals.file_delete_message_multiple_mixed_format
            }
            getQuantityString(messageRes, files.size, files.size)
        }
        return MaterialAlertDialogBuilder(requireContext(), theme)
            .setMessage(message)
            .setPositiveButton(android.R.string.ok) { _, _ -> listener.deleteFiles(files) }
            .setNegativeButton(android.R.string.cancel, null)
            .create()
    }

    companion object {
        fun show(files: FileItemSet, fragment: Fragment) {
            ConfirmDeleteFilesDialogFragment().putArgs(Args(files)).show(fragment)
        }
    }

    @Parcelize
    class Args(val files: FileItemSet) : ParcelableArgs

    interface Listener {
        fun deleteFiles(files: FileItemSet)
    }
}
